### ISSUE TITLE
Unexpected `Unknown` type arguments when using class factory which has access to bound and defaults

### ISSUE THREAD

**Issue by Daraan** [microsoft/pyright#10688]
**Describe the problem**
👋🙂
I have a factory method that fixes a class attribute in a dynamic subclass, further it should fix the types of the resulting subclass.
The argument to be passed has default values for its type parameters and all are bound.
In my case 2 out of 3 have default values but my problem appears also if all three have defaults.

**Code or Screenshots**

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAMoCmANsALABQjAxtQIYDOHhIaA2gCUAXIWJQAvJhQwy5EanySFZACoiOMEBKgaQAXSGMoxqGxiaR5IybOaATCJWNrxgAKtOHCJRgALMAAmLlABlMCwiJQBABSsHACUUAC0AHwRCJR8VLQGxgDEIQCuEBBwUMCFKMwwSGAoptwA7rgA1mzglUEMJj1QID6FIPVxzgyh4ZUIbMwtgiJEpBTy0qrqmnrR7Wgi8Bl8BLyCsqp6eompsIUI1JkCx1AquQB0L6MeXFAAYmBgAEKct3mxFkyxkDzW%2BkM3RsvCgO0i%2B0OdwoJz0wWC7nYXG8fkCwXG5WmMFwcD4ICBiw4oLIMAhGziZC2AH14XsDvwQGQODTTuc0rtMgAib5-AFkrk8wUGYI9d7cAAiYVQURF0RF-w4mU5Oh58ShvQNplhkmZMpM-Rgg3qiuAyoCItGPQJTKZqCQMBd0U1tD1ZuM3uAT1sICZVIo2gDQd40XiQfMID9OhogeDTNpDwjyajaBjcfs6KYhf6ADdKGxqGnItF2fFGCKAIzadUAp7AIkk6u8WsMYA-RuSBsxxglssVgXRXtgeuJEwFSgADwy1SiXx%2BGsBUGiCiguFEcHiIKkYLUOnWUJH5crGQnfbzIYS%2BSgC6XMBX25QYHwAFUUC0P40UGHShS0vcdJ3rO80xnOdF0oZcAlPLQP2-X9-0AhgCgAdVaVA0DrH47CbNcWzbaoO3ZfZiDORhJ0IgcCKHBgLzHKtaO7Zir0oG8wDsSCHygAoFCAkCWOvWjIJgaDEPwsAAGYiNFTVW3bEA4E7NAngFGJ4m7Sd5PouTGI4sCflkxICjQH4uiAA)

```python
from typing import Any, Self

class Arg[R: Any = int, S: int = int, T: str = str]:
    attr: S
    attr2: T

    @classmethod
    def typed(cls) -> type[Self]:  # dummy function as workaround
        return cls

def unpack[R: Any, S: int, T: str](arg: type[Arg[R, S, T]]) -> tuple[R, S, T]: ...

class FooBase[R: Any, S: int, T: str]:
    arg : type[Arg[R, S, T]]
    
    @classmethod
    def factory[r: Any, s: int, t: str](cls, arg_: type[Arg[r, s, t]]) -> type["FooBase[r, s, t]"]:
        class DefinedFoo(FooBase[r, s, t]):
            arg = arg_
        return DefinedFoo

    def __init__(self):
        self.attr_s: S = self.arg().attr
        self.attr_t: T = self.arg().attr2
    


reveal_type(Arg)
Foo1 = FooBase.factory(Arg)
foo1 = Foo1()
reveal_type(foo1)    # expected FooBase[R: (int or Any), S: int, T: str]:
reveal_type(foo1.attr_s)  # expected int not Unknown
reveal_type(foo1.attr_t)  # expected str not Unknown
# Working
Foo2 = FooBase.factory(Arg[Any])
foo2 = Foo2()
reveal_type(foo2)
reveal_type(foo2.attr_s)  # int
reveal_type(foo2.attr_t)  # str
Foo3 = FooBase.factory(Arg.typed())
foo3 = Foo3()
reveal_type(foo3) # good

```

**Describe the solution you’d like**

I would like to see `Foo` to be specified with the default values (or the bounds) the `Arg` when passing it to `factory`, even the bounds of the `FooBase` or `factory` would be fine. That the resulting class is `Unknown` in all three type parameters is unexpected and makes downstream usage more complicated, e.g. resulting in no type hint and type-completion (via Pylance) for `foo.attr_s.bar` when it is a more complex type.
I assume this behavior is because the binding rules are not triggered here.
I think it would be helpful if that were done here, to not need alternatives like `FooBase.factory(Arg[Any])` or `FooBase.factory(Arg.typed())` - a further alternative is to define defaults for the type parameters of `factory` but that seems unsafe as that would disregard the `bound/default` of the actual argument.

**VS Code extension or command-line**
1.1.403 (playground)

**Comment by erictraut** [comment]
Yeah, this looks like a bug. The subexpression `Arg` in the expression `FooBase.factory(Arg)` should be evaluated as `Arg[int, int, str]`, but it appears to be treated as `Arg[Unknown, Unknown, Unknown]`.
